source/iterator/cdc_iterator.go

```diff
@@ -0,0 +1,269 @@
+	logger.Trace().Msg("Starting the flushLatestLeads")
+	token, err := c.client.GetNextPageToken(c.lastModified)
+	c.lastModified = time.Now().UTC()
+	if err != nil {
+		logger.Error().Err(err).Msg("Error while getting the next page token")
+		return err
+	}
+	changedLeadIds, err := c.GetChangedLeadsIDs(ctx, token)
+	if err != nil {
+		logger.Error().Err(err).Msg("Error while getting the changed leads")
+		return err
+	}
+	deletedLeadIds, err := c.GetDeletedLeadsIDs(ctx, token)
+	if err != nil {
+		logger.Error().Err(err).Msg("Error while getting the deleted leads")
+		return err
+	}
+	for _, id := range deletedLeadIds {
+		c.buffer <- Record{
+			id:      id,
+			deleted: true,
+			data:    nil,
+		}
+	}
+	for _, id := range changedLeadIds {
+		res, err := c.client.GetLeadByID(id, c.fields)
+		if err != nil {
+			logger.Error().Err(err).Msg("Error while getting the lead")
+			return err
+		}
+		dataMap := make([]map[string]interface{}, 0)
+		err = json.Unmarshal(*res, &dataMap)
+		if err != nil {
+			logger.Error().Err(err).Msg("Error while unmarshalling the lead")
+			return err
+		}
+		for _, data := range dataMap {
+			c.buffer <- Record{
+				id:      id,
+				deleted: false,
+				data:    data,
+			}
+		}
+	}
+	return nil
+}
+
+// returns list of deleted leads ids.
+func (c *CDCIterator) GetDeletedLeadsIDs(ctx context.Context, token string) ([]int, error) {
+	response, err := c.client.GetDeletedLeads(token)
+	if err != nil {
+		return nil, err
+	}
+	if len(*response) == 0 {
+		return nil, nil
+	}
+	var deletedLeadResults []map[string]interface{}
+	err = json.Unmarshal(*response, &deletedLeadResults)
+	if err != nil {
+		return nil, err
+	}
+	var leadIds = make([]int, 0)
+	for _, deletedLeadResult := range deletedLeadResults {
+		var id = int(deletedLeadResult["leadId"].(float64))
+		leadIds = append(leadIds, id)
+	}
+	return leadIds, nil
+}
+
+// returns list of changed leads ids.
+func (c *CDCIterator) GetChangedLeadsIDs(ctx context.Context, token string) ([]int, error) {
+	var leadIds = make(map[int]struct{}) // using map to avoid duplicates
+	moreResult := true
+	for moreResult {
+		response, err := c.client.GetLeadChanges(token, c.fields)
+		if err != nil {
+			return nil, err
+		}
+		if len(response.Result) == 0 {
+			return nil, nil
+		}
+		moreResult = response.MoreResult
+		token = response.NextPageToken
+		var leadChangeResults []map[string]interface{}
+		err = json.Unmarshal(response.Result, &leadChangeResults)
+		if err != nil {
+			return nil, err
+		}
+		for _, leadChangeResult := range leadChangeResults {
+			var activityTypeID = leadChangeResult["activityTypeId"].(float64)
+			if activityTypeID == ActivityTypeIDNewLead || activityTypeID == ActivityTypeIDChangeDataValue {
+				var id = int(leadChangeResult["leadId"].(float64))
+				leadIds[id] = struct{}{}
+			}
+		}
+	}
+	keys := make([]int, 0, len(leadIds))
+	for k := range leadIds {
+		keys = append(keys, k)
+	}
+	sort.Ints(keys)
```

Comment:
actually initial implementaion required, now it is not, i removed.

Reply:
it requires for some edge cases, especially for `func (a acceptanceTest) TestSource_Open_ResumeAtPositionCDC(t *testing.T) {` tests, sorting helps in choosing the last key and resuming the CDC mode

Reply:
So the data from Marketo is not sorted?

Reply:
a kind of no, coz we get records based on change time, it can be any record. For eg: at given time if  id=1234 updated first and later id=1000 updated, then we get records in sequence 1234,1000 so on.

Reply:
So why is that a problem? The position is based on the change time, so the next time CDC starts, it starts from the last processed change time.

Reply:
actually without sorting, connector will work fine, but some tests in acceptance tests are failing. especially for `func (a acceptanceTest) TestSource_Open_ResumeAtPositionCDC(t *testing.T) {` tests, sorting helps in choosing the last key and resuming the CDC mode

Reply:
> So why is that a problem? The position is based on the change time, so the next time CDC starts, it starts from the last processed change time.

in this `func (a acceptanceTest) TestSource_Open_ResumeAtPositionCDC(t *testing.T) { ` test we genrate 10 rec and ack for 5. And then expect other 5 recs. But all these 10 rec are updated in same time. so we'll get same 10 recs again. So we use `lastKey` to continue from where we left.

Reply:
@rustiever So the keys sorted, are lead IDs. Are lead IDs actually timestamps? I might be missing something, but I can't see a connection. We use the last modified timestamp of a lead in the position, and when resuming CDC. But here, we are sorting lead IDs to make sure we can resume from a position.

Reply:
we primarily use `lastModiefiedTime` for CDC. But sometimes this doesn't work when we resume from CDC. For eg: at 10:15 time we get 8 recs 5,6,3,2,7,1,8,4. Lets assume we processed 4(5,6,3,2) records among them. timestamp for all records are same. So when we resume CDC, We get again same set of records. we need a way to filter the processed records. So we use sorting. once the records are sorted such as 1,2,3,4,5,6,7,8, and CDC is resumed we get lasttime and lastKey. We compare fetched records with lastKey to filter which records are not processed.